### ISSUE TITLE
False positive for unnecessaryDefer

### ISSUE THREAD

**Issue by bombsimon** [go-critic/go-critic#926]
It seems like the check for `unnecessaryDefer` only checks if the defer statement is the last statement in a block, no matter what block it is. Because of this it's not possible to set conditional defer statements.

Noticed this when a build in [`golangci-lint`](https://github.com/golangci/golangci-lint/pull/1148) failed. Here's a smaller example to reproduce.

```go
package main

import "fmt"

func main() {
    if true {
        defer func() {
            fmt.Println("something was true")
        }()
    }

    fmt.Println("more code before return")
}
```

```sh
$ gocritic check -enable='unnecessaryDefer' ./...
./main.go:7:3: unnecessaryDefer: defer func(){...}(...) is placed just before return
```

**Comment by quasilyte** [comment]
Fixed in #927
